File: gps_agent_pkg/src/positioncontroller.cpp
#include "gps_agent_pkg/util.h"
#include "gps_agent_pkg/robotplugin.h"
#include "gps_agent_pkg/superchickplugin.h"
#include "gps_agent_pkg/positioncontroller.h"

using namespace gps_control;

// Constructor.

// Constructor.
PositionController::PositionController(ros::NodeHandle& n, gps::ActuatorType arm, int size)
    : Controller(n, arm, size)
{
    // Initialize velocity bounds.
    max_velocities_.resize(size);

    // Initialize current angle and position.
    current_angles_.resize(size);
    current_angle_velocities_.resize(size);
    current_pose_.resize(size);

    // Initialize target angle and position.
    target_angles_.resize(size);
    target_pose_.resize(size);

    // Initialize joints temporary storage.
    temp_angles_.resize(size);

    // Set initial mode.
    mode_ = gps::NO_CONTROL;

    // Set initial time.
    last_update_time_ = ros::Time(0.0);

    // Set arm.
    arm_ = arm;

    //
    report_waiting = false;


    //instantiate camera sensor class
    camerasensor = new CameraSensor; //(n, &robotplugin);
}

// Destructor.
PositionController::~PositionController()
{
}


// Update the controller (take an action).
void PositionController::update(RobotPlugin *plugin, ros::Time current_time, boost::scoped_ptr<Sample>& sample, Eigen::VectorXd &torques)
{
/*    // Get current joint angles.
    plugin->get_task_space_readings(temp_angles_, arm_);

    // Check dimensionality.
    assert(temp_angles_.rows() == torques.rows());
    assert(temp_angles_.rows() == current_angles_.rows());

    // Estimate joint angle velocities.
    double update_time = current_time.toSec() - last_update_time_.toSec();
    if (!last_update_time_.isZero())
    { // Only compute velocities if we have a previous sample.
        current_angle_velocities_ = (temp_angles_ - current_angles_)/update_time;
    }

    // Store new angles.
    current_angles_ = temp_angles_;

    // Update last update time.
    last_update_time_ = current_time;*/

    // If doing task space control, compute task positions target.
    if (mode_ == gps::TASK_SPACE)
    {
        
        // TODO: implement.

        // Get current end effector position. latest_vicon_pose_
        geometry_msgs::Twist current_headpose = camerasensor->latest_vicon_pose_[1];
        //retrieve the markers just for a check
        geometry_msgs::Point fore  = latest_vicon_markers_[0];
        geometry_msgs::Point left  = latest_vicon_markers_[1];
        geometry_msgs::Point right = latest_vicon_markers_[2];
        geometry_msgs::Point chin  = latest_vicon_markers_[3];
        // Get current Jacobian.

        // estimate twist, i.e. task space velocities
        double update_time = current_time.toSec() - last_update_time_.toSec();
        if (!last_update_time_.isZero())
        { // Only compute velocities if we have a previous sample.
            current_angle_velocities_ = (temp_angles_ - current_angles_)/update_time;
        }

        // Store new angles.
        current_angles_ = temp_angles_;

        // Update last update time.
        last_update_time_ = current_time;

        // TODO: should also try Jacobian pseudoinverse, it may work a little better.
        // Compute desired joint angle offset using Jacobian transpose method.
        target_angles_ = current_angles_ + temp_jacobian_.transpose() * (target_pose_ - current_pose_);
    }

    // If we're doing any kind of control at all, compute torques now.
    if (mode_ != gps::NO_CONTROL)
    {
        // Compute error.
        temp_angles_ = current_angles_ - target_angles_;

        // Add to integral term.
        pd_integral_ += temp_angles_ * update_time;

        // Clamp integral term
        for (int i = 0; i < temp_angles_.rows(); i++){
            if (pd_integral_(i) * pd_gains_i_(i) > i_clamp_(i)) {
                pd_integral_(i) = i_clamp_(i) / pd_gains_i_(i);
            }
            else if (pd_integral_(i) * pd_gains_i_(i) < -i_clamp_(i)) {
                pd_integral_(i) = -i_clamp_(i) / pd_gains_i_(i);
            }
        }

        // Compute torques.
        torques = -((pd_gains_p_.array() * temp_angles_.array()) +
                    (pd_gains_d_.array() * current_angle_velocities_.array()) +
                    (pd_gains_i_.array() * pd_integral_.array())).matrix();
    }
    else
    {
        torques = Eigen::VectorXd::Zero(torques.rows());
    }

}

// Configure the controller.
void PositionController::configure_controller(OptionsMap &options)
{
    // This sets the target position.
    // This sets the mode
    ROS_INFO_STREAM("Received controller configuration");
    // needs to report when finished
    report_waiting = true;
    mode_ = (gps::PositionControlMode) boost::get<int>(options["mode"]);
    if (mode_ != gps::NO_CONTROL){
        Eigen::VectorXd data = boost::get<Eigen::VectorXd>(options["data"]);
        Eigen::MatrixXd pd_gains = boost::get<Eigen::MatrixXd>(options["pd_gains"]);
        for(int i=0; i<pd_gains.rows(); i++){
            pd_gains_p_(i) = pd_gains(i, 0);
            pd_gains_i_(i) = pd_gains(i, 1);
            pd_gains_d_(i) = pd_gains(i, 2);
            i_clamp_(i) = pd_gains(i, 3);
        }
        if(mode_ == gps::JOINT_SPACE){
            target_angles_ = data;
        }else{
            ROS_ERROR("Unimplemented position control mode!");
        }
    }
}

// Check if controller is finished with its current task.
bool PositionController::is_finished() const
{
    // Check whether we are close enough to the current target.
    if (mode_ == gps::JOINT_SPACE){
        double epspos = 0.185;
        double epsvel = 0.01;
        double error = (current_angles_ - target_angles_).norm();
        double vel = current_angle_velocities_.norm();
        return (error < epspos && vel < epsvel);
    }
    else if (mode_ == gps::NO_CONTROL){
        return true;
    }
}

// Reset the controller -- this is typically called when the controller is turned on.
void PositionController::reset(ros::Time time)
{
    // Clear the integral term.
    pd_integral_.fill(0.0);

    // Clear update time.
    last_update_time_ = ros::Time(0.0);
}

File: python/gps_agent_pkg/position_controller.py
"""
Translates gps_agent_pkg/src/positioncontroller.cpp + include/positioncontroller.h.

PID joint-space position controller.
is_finished() threshold: position error < 0.185 rad AND velocity < 0.01 rad/s.
"""
from __future__ import annotations

from typing import Optional

import numpy as np

from gps_agent_pkg.controller import Controller
from gps_agent_pkg.sample import ControllerSample

# gps.proto PositionControlMode values
NO_CONTROL  = 0
JOINT_SPACE = 1
TASK_SPACE  = 2


class PositionController(Controller):
    """
    PD+I joint-space position controller.

    Mirrors C++ gps_control::PositionController.
    """

    EPS_POS: float = 0.185   # rad — position convergence threshold
    EPS_VEL: float = 0.01    # rad/s — velocity convergence threshold

    def __init__(self, node, arm: int, n_joints: int) -> None:
        super().__init__()
        self._arm = arm
        self._n_joints = n_joints

        self._mode: int = NO_CONTROL
        self.report_waiting: bool = False

        self._pd_gains_p = np.zeros(n_joints, dtype=np.float64)
        self._pd_gains_i = np.zeros(n_joints, dtype=np.float64)
        self._pd_gains_d = np.zeros(n_joints, dtype=np.float64)
        self._i_clamp    = np.zeros(n_joints, dtype=np.float64)

        self._pd_integral            = np.zeros(n_joints, dtype=np.float64)
        self._current_angles         = np.zeros(n_joints, dtype=np.float64)
        self._current_angle_velocities = np.zeros(n_joints, dtype=np.float64)
        self._target_angles          = np.zeros(n_joints, dtype=np.float64)

        self._last_update_time: Optional[float] = None

    # ------------------------------------------------------------------
    # Configuration
    # ------------------------------------------------------------------

    def configure_controller(self, options: dict) -> None:
        """
        Parse mode, PD gains, and target angles.
        Mirrors C++ PositionController::configure_controller.
        """
        self.report_waiting = True
        self._mode = int(options["mode"])

        if self._mode != NO_CONTROL:
            pd_gains = np.asarray(options["pd_gains"], dtype=np.float64)
            # pd_gains has shape (n_joints, 4): [Kp, Ki, Kd, i_clamp]
            self._pd_gains_p = pd_gains[:, 0].copy()
            self._pd_gains_i = pd_gains[:, 1].copy()
            self._pd_gains_d = pd_gains[:, 2].copy()
            self._i_clamp    = pd_gains[:, 3].copy()

            if self._mode == JOINT_SPACE:
                self._target_angles = np.asarray(options["data"], dtype=np.float64).copy()
            # TASK_SPACE not yet implemented (mirrors C++ TODO)

    # ------------------------------------------------------------------
    # Update
    # ------------------------------------------------------------------

    def update(
        self,
        plugin,
        current_time: float,
        sample: ControllerSample,
        torques: np.ndarray,
    ) -> np.ndarray:
        """
        One PID step: read joint angles, compute torques.
        Mirrors C++ PositionController::update.
        """
        # Get current joint angles from plugin
        q = plugin.get_joint_encoder_readings(self._arm)

        # Estimate velocities
        dt = 0.0
        if self._last_update_time is not None:
            dt = current_time - self._last_update_time
        if dt > 0:
            self._current_angle_velocities = (q - self._current_angles) / dt

        self._current_angles = q.copy()
        self._last_update_time = current_time

        if self._mode != NO_CONTROL:
            err = self._current_angles - self._target_angles

            # Accumulate and clamp integral term
            self._pd_integral += err * dt
            safe_ki = np.where(self._pd_gains_i != 0, self._pd_gains_i, 1.0)
            np.clip(
                self._pd_integral,
                -self._i_clamp / safe_ki,
                 self._i_clamp / safe_ki,
                out=self._pd_integral,
            )

            torques[:] = -(
                self._pd_gains_p * err
                + self._pd_gains_d * self._current_angle_velocities
                + self._pd_gains_i * self._pd_integral
            )
        else:
            torques[:] = 0.0

        return torques

    # ------------------------------------------------------------------
    # Status
    # ------------------------------------------------------------------

    def is_finished(self) -> bool:
        """
        Return True when position error < EPS_POS AND velocity < EPS_VEL.
        Mirrors C++ PositionController::is_finished.
        """
        if self._mode == JOINT_SPACE:
            err = np.linalg.norm(self._current_angles - self._target_angles)
            vel = np.linalg.norm(self._current_angle_velocities)
            return bool(err < self.EPS_POS and vel < self.EPS_VEL)
        return self._mode == NO_CONTROL

    def reset(self, current_time: float) -> None:
        self._pd_integral[:] = 0.0
        self._last_update_time = None
